From 7323076f426cd9f4a94ee68475a8662abe99eaef Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
Date: Tue, 11 Aug 2015 00:47:29 +0300
Subject: [PATCH 2040/2066] v4l: vsp1: Reset the DRM pipeline when stopping
 the stream

When the stream is stopped all inputs need to be reset to NULL in order
to make them configurable again when restarting the stream. Don't use
vsp1_pipeline_reset() for that purpose as it would reset the whole
pipeline including the entities list that have been setup at init time.

Signed-off-by: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
[zou:Original patch taken from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git
 v4.9/rcar-3.5.]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 drivers/media/platform/vsp1/vsp1_drm.c |   10 ++++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/vsp1/vsp1_drm.c b/drivers/media/platform/vsp1/vsp1_drm.c
index 20d7d28..df86abb 100644
--- a/drivers/media/platform/vsp1/vsp1_drm.c
+++ b/drivers/media/platform/vsp1/vsp1_drm.c
@@ -43,7 +43,7 @@ static int vsp1_drm_pipeline_run(struct vsp1_pipeline *pipe)
 				struct vsp1_rwpf *rpf =
 					to_rwpf(&entity->subdev);
 
-				if (!rpf->fmtinfo)
+				if (!pipe->inputs[rpf->entity.index])
 					continue;
 			}
 
@@ -133,6 +133,13 @@ int vsp1_du_setup_lif(struct device *dev, unsigned int width,
 
 		media_entity_pipeline_stop(&pipe->output->entity.subdev.entity);
 
+		for (i = 0; i < bru->entity.source_pad; ++i) {
+			bru->inputs[i].rpf = NULL;
+			pipe->inputs[i] = NULL;
+		}
+
+		pipe->num_inputs = 0;
+
 		vsp1_device_put(vsp1);
 
 		dev_dbg(vsp1->dev, "%s: pipeline disabled\n", __func__);
@@ -306,7 +313,6 @@ int vsp1_du_setup_rpf(struct device *dev, unsigned int rpf_index,
 		/* Remove the RPF from the pipeline. */
 		vsp1->bru->inputs[rpf_index].rpf = NULL;
 		pipe->inputs[rpf_index] = NULL;
-		rpf->fmtinfo = NULL;
 
 		vsp1->drm->update = true;
 		start_stop = --pipe->num_inputs == 0;
-- 
1.7.5.4

